sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads.py

```diff
@@ -9,20 +9,62 @@
 from itertools import islice
 from math import ceil
 from threading import Lock
-from typing import Optional, Union
+from typing import AnyStr, IO, Iterable, Optional, Tuple, Union
 
 from azure.core.tracing.common import with_current_context
 
 from . import encode_base64, url_quote
-from .request_handlers import get_length
+from .request_handlers import get_length, read_length
 from .response_handlers import return_response_headers
-from .validation import calculate_crc64, calculate_md5, ChecksumAlgorithm
+from .streams import StructuredMessageEncodeStream, StructuredMessageProperties
+from .validation import calculate_crc64_bytes, calculate_md5, ChecksumAlgorithm
 
 
 _LARGE_BLOB_UPLOAD_MAX_READ_BUFFER_SIZE = 4 * 1024 * 1024
 _ERROR_VALUE_SHOULD_BE_SEEKABLE_STREAM = "{0} should be a seekable file-like/io.IOBase type stream object."
 
 
+def prepare_upload_data(
+    data: Union[bytes, str, IO[bytes], Iterable[AnyStr]],
+    encoding: str,
+    data_length: Optional[int],
+    validate_content: Union[bool, str, None]
+) -> Tuple[Union[bytes, IO[bytes]], int, int]:
+    # Trim the incoming data per provided length
+    if data_length is not None and isinstance(data, (str, bytes)) and data_length != len(data):
+        data = data[:data_length]
+    # Encode raw string data
+    if isinstance(data, str):
+        data = data.encode(encoding)
+
+    # Attempt to determine length of data if it's not provided
+    if data_length is None:
+        data_length = get_length(data)
+    # If we still can't get the length, read all the data into memory
+    if data_length is None:
+        data_length, data = read_length(data)
+
+    structured_message = validate_content == ChecksumAlgorithm.CRC64
+    if isinstance(data, bytes):
+        # Structured message requires a stream
+        if structured_message:
+            data = BytesIO(data)
+    elif hasattr(data, 'read'):
+        # TODO: Block IO[str] here?
+        pass
+    elif hasattr(data, '__iter__') and not isinstance(data, (list, tuple, set, dict)):
+        data = IterStreamer(data, encoding=encoding)
+    else:
+        raise TypeError(f"Unsupported data type: {type(data)}")
+
+    content_length = data_length
+    if structured_message:
+        data = StructuredMessageEncodeStream(data, data_length, StructuredMessageProperties.CRC64)
+        content_length = len(data)
```

Comment:
nit: Do we need this here? It seems that the all the work above to determine the length (i.e. validating if provided length is valid or even just reading all the data into memory and then getting the length) means that line 63 should always match line 60?

Reply:
Ah, yeah, notice we return two different lengths below. In most cases these are equal but for structured message the content length will end up being larger than the data length (due to extra headers and such). Calling `len` on the `StructuredMessageEncodeStream` will give us the full length of the stream (i.e. content length) which will be different than any length we have calculated prior.